compiler/src/iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.cpp

```diff
@@ -0,0 +1,266 @@
+                        VectorLayoutOptions &options,
+                        PatternBenefit benefit = 1)
+      : OpTraitRewritePattern<OpTrait::Elementwise>(context, benefit),
+        analysis(analysis), options(options) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    if (!OpTrait::hasElementwiseMappableTraits(op))
+      return failure();
+
+    // Get the distributed operands.
+    SmallVector<Value> operands;
+    for (Value operand : op->getOperands()) {
+      if (auto vectorOperand = dyn_cast<VectorValue>(operand)) {
+        operand = getDistributed(rewriter, vectorOperand, options);
+      }
+      operands.push_back(operand);
+    }
+
+    // Get the new distributed vector types for the operation.
+    SmallVector<Type> resultTypes;
+    for (Value result : op->getResults()) {
+      Type resultType = result.getType();
+
+      // Distribute vector result types.
+      if (auto vectorResult = dyn_cast<VectorValue>(result)) {
+        resultType = VectorType::get(options.getDistributedShape(vectorResult),
+                                     vectorResult.getType().getElementType());
+      }
+      resultTypes.push_back(resultType);
+    }
+
+    // Replace the original op with the distributed op.
+    Operation *distributedOp =
+        rewriter.create(op->getLoc(), op->getName().getIdentifier(), operands,
+                        resultTypes, op->getAttrs());
+    replaceOpWithDistributedValues(rewriter, op, options,
+                                   distributedOp->getResults());
+    return success();
+  }
+
+private:
+  VectorLayoutAnalysis &analysis;
+  VectorLayoutOptions &options;
+};
+
+VectorDistribution::VectorDistribution(func::FuncOp root,
+                                       VectorLayoutAnalysis &analysis,
+                                       VectorLayoutOptions &options)
+    : root(root), analysis(analysis), options(options) {
+  options.setAnchorOps();
+  if (failed(analysis.run()))
+    return;
+  LLVM_DEBUG(llvm::dbgs() << "Layout Analysis Completed Successfully :\n");
+  LLVM_DEBUG(analysis.print(llvm::dbgs()));
+  LLVM_DEBUG(llvm::dbgs() << "\n\n");
+}
+
+/// A rewriter for the pattern rewriting driver.
+class VectorDistributionRewriter : public PatternRewriter,
+                                   RewriterBase::Listener {
+public:
+  VectorDistributionRewriter(MLIRContext *ctx, DenseSet<Operation *> &erasedOps,
+                             SmallVectorImpl<Operation *> &worklist)
+      : PatternRewriter(context, this), erasedOps(erasedOps),
+        worklist(worklist) {}
+
+  /// We need to keep track of erased operations so that we never try to
+  /// distribute them again.
+  void notifyOperationRemoved(Operation *op) override { erasedOps.insert(op); }
+
+private:
+  // Reference to operations to be erased and the worklist of the driver.
+  DenseSet<Operation *> &erasedOps;
+  SmallVectorImpl<Operation *> &worklist;
```

Comment:
I am not sure having the worklist in the rewriter is a good idea... I had left a comment earlier of moving the worklist out of the rewriter. Is there a precedence of the rewriter itself having a worklist that you are going off of?

Reply:
Removed the worklist from the rewriter.